File: src/main/kotlin/cz/uhk/ppro/mhjp/resftulgallery/rest/GalleryController.kt
package cz.uhk.ppro.mhjp.resftulgallery.rest

import cz.uhk.ppro.mhjp.resftulgallery.dto.AddRemoveImagesFromGalleryDto
import cz.uhk.ppro.mhjp.resftulgallery.dto.SubmitGalleryDto
import cz.uhk.ppro.mhjp.resftulgallery.service.GalleryService
import cz.uhk.ppro.mhjp.resftulgallery.util.HeaderUtil
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest

@RestController
class GalleryController(
        private val galleryService: GalleryService,
        private val headerUtil: HeaderUtil
) {

    @PostMapping("/api/g")
    fun createGallery(@RequestBody newGallery: SubmitGalleryDto, request: HttpServletRequest)
            = galleryService.createEntity(newGallery, headerUtil.getTokenFromHeader(request))

    @GetMapping("/api/g/{uuid}")
    fun retrieveGallery(@PathVariable("uuid") uuid: String, request: HttpServletRequest)
            = galleryService.readEntity(uuid, headerUtil.getOptionalTokenFromHeader(request))

    @PutMapping("/api/g/{uuid}")
    fun editGallery(@PathVariable("uuid") uuid: String, @RequestBody editGallery: SubmitGalleryDto, request: HttpServletRequest)
            = galleryService.updateEntity(uuid, editGallery, headerUtil.getTokenFromHeader(request))

    @DeleteMapping("/api/g/{uuid}")
    fun deleteGallery(@PathVariable("uuid") uuid: String, request: HttpServletRequest)
            = galleryService.deleteEntity(uuid, headerUtil.getTokenFromHeader(request))

    @PostMapping("/api/g/{uuid}/add")
    fun addImages(@PathVariable("uuid") uuid: String, @RequestBody add: AddRemoveImagesFromGalleryDto, request: HttpServletRequest)
            = galleryService.addImages(uuid, add, headerUtil.getTokenFromHeader(request))

    @DeleteMapping("/api/g/{uuid}/remove")
    fun removeImages(@PathVariable("uuid") uuid: String, @RequestBody remove: AddRemoveImagesFromGalleryDto, request: HttpServletRequest)
            = galleryService.removeImages(uuid, remove, headerUtil.getTokenFromHeader(request))

}

File: src/main/bobril/cz/uhk/ppro/mhjp/gallery/api/rest/GalleryController.ts
import {ErrorResponseDto, SuccessResponseDto} from "../dto/baseDto";
import {httpDelete, httpGet, httpPost, httpPut} from "../http";
import {AddRemoveImagesFromGalleryDto, GalleryDataDto, SubmitGalleryDto} from "../dto/galleryDto";

export const createGallery =  (newGallery: SubmitGalleryDto,
                            success?: (response: GalleryDataDto, status: number, headers: (key: string) => string) => void,
                                 error?: (response: GalleryDataDto, status: number) => void,) => {
    httpPost("/api/g/", newGallery, success, error);
};

export const retrieveGallery =  (uuid: string,
                               success?: (response: GalleryDataDto, status: number, headers: (key: string) => string) => void,
                               error?: (response: GalleryDataDto, status: number) => void,) => {
    httpGet("/api/g/" + uuid, undefined, success, error);
};

export const editGallery =  (uuid: string,
                                 editGallery: SubmitGalleryDto,
                                 success?: (response: GalleryDataDto, status: number, headers: (key: string) => string) => void,
                                 error?: (response: GalleryDataDto, status: number) => void,) => {
    httpPut("/api/g/" + uuid, editGallery, success, error);
};
export const addImages =  (uuid: string,
                             add: AddRemoveImagesFromGalleryDto,
                             success?: (response: SuccessResponseDto, status: number, headers: (key: string) => string) => void,
                             error?: (response: ErrorResponseDto, status: number) => void,) => {
    httpPost("/api/g/"+uuid+"/add" + uuid, add, success, error);
};
export const deleteGallery =  (uuid: string,
                               remove: AddRemoveImagesFromGalleryDto,
                               success?: (response: SuccessResponseDto, status: number, headers: (key: string) => string) => void,
                               error?: (response: ErrorResponseDto, status: number) => void,) => {
    httpDelete("/api/g/" + uuid + "/remove", undefined, success, error);
};